File: tests/reporting_utils.py
#!/usr/bin/env python

from random import randint
from ethereum.tools import tester
from ethereum.tools.tester import TransactionFailed
from pytest import fixture, mark, raises
from datetime import timedelta
from utils import bytesToHexString, longToHexString, PrintGasUsed, TokenDelta, EtherDelta

def proceedToDesignatedReporting(fixture, market):
    fixture.contracts["Time"].setTimestamp(market.getEndTime() + 1)

def proceedToInitialReporting(fixture, market):
    fixture.contracts["Time"].setTimestamp(market.getDesignatedReportingEndTime() + 1)

def proceedToNextRound(fixture, market, contributor = tester.k0, doGenerateFees = False, moveTimeForward = True, randomPayoutNumerators = False, description = ""):
    if fixture.contracts["Controller"].getTimestamp() < market.getEndTime():
        fixture.contracts["Time"].setTimestamp(market.getDesignatedReportingEndTime() + 1)

    disputeWindow = market.getDisputeWindow()

    payoutNumerators = [0] * market.getNumberOfOutcomes()
    payoutNumerators[1] = market.getNumTicks()

    if (disputeWindow == longToHexString(0)):
        market.doInitialReport(payoutNumerators, "")
        assert market.getDisputeWindow()
    else:
        disputeWindow = fixture.applySignature('DisputeWindow', market.getDisputeWindow())
        if market.getDisputePacingOn():
            fixture.contracts["Time"].setTimestamp(disputeWindow.getStartTime() + 1)
        # This will also use the InitialReporter which is not a DisputeCrowdsourcer, but has the called function from abstract inheritance
        winningReport = fixture.applySignature('DisputeCrowdsourcer', market.getWinningReportingParticipant())
        winningPayoutHash = winningReport.getPayoutDistributionHash()

        chosenPayoutNumerators = [0] * market.getNumberOfOutcomes()
        chosenPayoutNumerators[1] = market.getNumTicks()

        if (randomPayoutNumerators):
            chosenPayoutNumerators[1] = randint(0, market.getNumTicks())
            chosenPayoutNumerators[2] = market.getNumTicks() - chosenPayoutNumerators[1]
        else:
            firstReportWinning = market.derivePayoutDistributionHash(payoutNumerators) == winningPayoutHash
            if firstReportWinning:
                chosenPayoutNumerators[1] = 0
                chosenPayoutNumerators[2] = market.getNumTicks()

        chosenPayoutHash = market.derivePayoutDistributionHash(chosenPayoutNumerators)
        amount = 2 * market.getParticipantStake() - 3 * market.getStakeInOutcome(chosenPayoutHash)
        with PrintGasUsed(fixture, "Contribute:", 0):
            market.contribute(chosenPayoutNumerators, amount, description, sender=contributor)
        assert market.getForkingMarket() or market.getDisputeWindow() != disputeWindow

    if (doGenerateFees):
        universe = fixture.applySignature("Universe", market.getUniverse())
        generateFees(fixture, universe, market)

    if (moveTimeForward):
        disputeWindow = fixture.applySignature('DisputeWindow', market.getDisputeWindow())
        fixture.contracts["Time"].setTimestamp(disputeWindow.getStartTime() + 1)

def proceedToFork(fixture, market, universe):
    while (market.getForkingMarket() == longToHexString(0)):
        proceedToNextRound(fixture, market)

    for i in range(market.getNumParticipants()):
        reportingParticipant = fixture.applySignature("DisputeCrowdsourcer", market.getReportingParticipant(i))
        reportingParticipant.forkAndRedeem()

def finalizeFork(fixture, market, universe, finalizeByMigration = True):
    reputationToken = fixture.applySignature('ReputationToken', universe.getReputationToken())

    # The universe forks and there is now a universe where NO and YES are the respective outcomes of each
    noPayoutNumerators = [0] * market.getNumberOfOutcomes()
    noPayoutNumerators[1] = market.getNumTicks()
    yesPayoutNumerators = [0] * market.getNumberOfOutcomes()
    yesPayoutNumerators[2] = market.getNumTicks()
    noUniverse =  fixture.applySignature('Universe', universe.createChildUniverse(noPayoutNumerators))
    yesUniverse =  fixture.applySignature('Universe', universe.createChildUniverse(yesPayoutNumerators))
    noUniverseReputationToken = fixture.applySignature('ReputationToken', noUniverse.getReputationToken())
    yesUniverseReputationToken = fixture.applySignature('ReputationToken', yesUniverse.getReputationToken())
    assert noUniverse.address != universe.address
    assert yesUniverse.address != universe.address
    assert yesUniverse.address != noUniverse.address
    assert noUniverseReputationToken.address != yesUniverseReputationToken.address

    # Attempting to finalize the fork now will not succeed as no REP has been migrated and not enough time has passed
    with raises(TransactionFailed):
        market.finalizeFork()

    # A Tester moves some of their REP to the YES universe
    amount = 10 ** 6 * 10 ** 18

    with raises(TransactionFailed):
        reputationToken.migrateOut(yesUniverseReputationToken.address, 0)

    with TokenDelta(yesUniverseReputationToken, amount, tester.a0, "Yes REP token balance was no correct"):
        reputationToken.migrateOut(yesUniverseReputationToken.address, amount)

    # Attempting to finalize the fork now will not succeed as a majority or REP has not yet migrated and fork end time has not been reached
    with raises(TransactionFailed):
        market.finalizeFork()

    if (finalizeByMigration):
        # Tester 0 moves more than 50% of REP
        amount = reputationToken.balanceOf(tester.a0) - 20
        with TokenDelta(noUniverseReputationToken, amount, tester.a0, "No REP token balance was no correct"):
            reputationToken.migrateOut(noUniverseReputationToken.address, amount)
        assert reputationToken.balanceOf(tester.a0) == 20
        assert market.getWinningPayoutDistributionHash() == noUniverse.getParentPayoutDistributionHash()
    else:
        # Time marches on past the fork end time
        fixture.contracts["Time"].setTimestamp(universe.getForkEndTime() + 1)
        assert market.finalize()
        assert market.getWinningPayoutDistributionHash() == yesUniverse.getParentPayoutDistributionHash()
        # If the fork is past the fork period we can not migrate
        with raises(TransactionFailed):
            reputationToken.migrateOut(yesUniverseReputationToken.address, 1)

    # Finalize fork cannot be called again
    with raises(TransactionFailed):
        market.finalizeFork()

def generateFees(fixture, universe, market):
    completeSets = fixture.contracts['CompleteSets']
    cash = fixture.contracts['Cash']
    mailbox = fixture.applySignature('Mailbox', market.getMarketCreatorMailbox())
    assert mailbox.withdrawEther()
    oldFeesBalance = cash.balanceOf(universe.getAuction())

    cost = 1000 * market.getNumTicks()
    marketCreatorFees = cost / market.getMarketCreatorSettlementFeeDivisor()
    completeSets.publicBuyCompleteSets(market.address, 1000, sender = tester.k1, value = cost)
    with TokenDelta(cash, marketCreatorFees, mailbox.address, "The market creator mailbox didn't get their share of fees from complete set sale"):
        completeSets.publicSellCompleteSets(market.address, 1000, sender=tester.k1)
    with EtherDelta(marketCreatorFees, market.getOwner(), fixture.chain, "The market creator did not get their fees when withdrawing ETH from the mailbox"):
        assert mailbox.withdrawEther()
    newFeesBalance = cash.balanceOf(universe.getAuction())
    reporterFees = cost / universe.getOrCacheReportingFeeDivisor()
    feesGenerated = newFeesBalance - oldFeesBalance
    assert feesGenerated == reporterFees, "Cash balance of auction higher by: " + str(fees - reporterFees)

File: source/tests-integration/ReportingUtils.ts
import BN = require('bn.js');
import { expect } from "chai";
import { TestFixture } from './TestFixture';
import { Market } from '../libraries/ContractInterfaces';

const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";

export class ReportingUtils {
    public async proceedToDesignatedReporting(fixture: TestFixture, market: Market) {
        const marketEndTime = await market.getEndTime_();
        await fixture.setTimestamp(marketEndTime.add(new BN(1)));
    }

    public async proceedToInitialReporting(fixture: TestFixture, market: Market) {
        const designatedReportingEndTime = await market.getDesignatedReportingEndTime_();
        await fixture.setTimestamp(designatedReportingEndTime.add(new BN(1)));
    }

    // TODO: Add `contributor` param, like `proceedToNextRound` function in Python tests
    public async proceedToNextRound(fixture: TestFixture, market: Market, doGenerateFees: boolean = false, moveTimeForward: boolean = true, randomPayoutNumerators: boolean = false) {
        const currentTimestamp = await fixture.getTimestamp();
        const marketEndTime = await market.getEndTime_();
        if (currentTimestamp.lt(marketEndTime)) {
            const marketDesignatedReportingEndTime = await market.getDesignatedReportingEndTime_();
            await fixture.setTimestamp(marketDesignatedReportingEndTime.add(new BN(1)));
        }

        const disputeWindowAddress = await market.getDisputeWindow_();

        const numberOfOutcomes = await market.getNumberOfOutcomes_();
        const numTicks = await market.getNumTicks_();
        let payoutNumerators = new Array(numberOfOutcomes.toNumber()).fill(new BN(0));
        payoutNumerators[1] = numTicks;

        let winningPayoutHash = "";
        if (disputeWindowAddress === ZERO_ADDRESS) {
            await market.doInitialReport(payoutNumerators, "");
            expect(await market.getDisputeWindow_() === ZERO_ADDRESS).to.be.false;
            console.log("Submitted initial report");

            // Buy and sell complete sets to generate reporting fees
            let outcome = new BN(0);
            let numShares = new BN(10000000000000);
            // Buy complete sets
            await fixture.buyCompleteSets(market, numShares);
            let numOwnedShares = await fixture.getNumSharesInMarket(market, outcome);

            let ethBalance = await fixture.getEthBalance();
            console.log("ethBalance after buying complete set", ethBalance.toString(10));
            console.log("numOwnedShares after buying complete set", numOwnedShares.toString(10));

            // Sell Complete Sets
            ethBalance = await fixture.getEthBalance();
            console.log("ethBalance before selling complete set", ethBalance.toString(10));
            let numOwnedSharesBefore = await fixture.getNumSharesInMarket(market, outcome);
            console.log("numOwnedShares before selling complete set", numOwnedSharesBefore.toString(10));
            await fixture.sellCompleteSets(market, numShares);
            ethBalance = await fixture.getEthBalance();
            console.log("ethBalance before selling complete set", ethBalance.toString(10));
            numOwnedSharesBefore = await fixture.getNumSharesInMarket(market, outcome);
            console.log("numOwnedShares after selling complete set", numOwnedSharesBefore.toString(10));
        } else {
            const disputeWindow = await fixture.getDisputeWindow(market);
            const disputeWindowStartTime = await disputeWindow.getStartTime_();
            await fixture.setTimestamp(disputeWindowStartTime.add(new BN(1)));
            // This will also use the InitialReporter which is not a DisputeCrowdsourcer, but has the called function from abstract inheritance
            const winningReport = await fixture.getWinningReportingParticipant(market);
            winningPayoutHash = await winningReport.getPayoutDistributionHash_();

            let chosenPayoutNumerators = new Array(numberOfOutcomes.toNumber()).fill(new BN(0));
            chosenPayoutNumerators[1] = numTicks;
            if (randomPayoutNumerators) {
                // Set chosenPayoutNumerators[1] to number >= 0 and <= numTicks
                chosenPayoutNumerators[1] = new BN(Math.floor(Math.random() * Math.floor(numTicks.toNumber() + 1)));
                chosenPayoutNumerators[2] = numTicks.sub(chosenPayoutNumerators[1]);
            } else {
                const firstReportWinning = await market.derivePayoutDistributionHash_(payoutNumerators) === winningPayoutHash;
                if (firstReportWinning) {
                    chosenPayoutNumerators[1] = new BN(0);
                    chosenPayoutNumerators[2] = numTicks;
                }
            }

            console.log("Reporting with payout numerators: ", chosenPayoutNumerators);
            const chosenPayoutHash = await market.derivePayoutDistributionHash_(chosenPayoutNumerators);
            const participantStake = await market.getParticipantStake_();
            const stakeInOutcome = await market.getStakeInOutcome_(chosenPayoutHash);
            const amount = participantStake.mul(new BN(2)).sub(stakeInOutcome.mul(new BN(3)));
            await fixture.contribute(market, chosenPayoutNumerators, amount);
            console.log("Staked", amount.toString(10));
            const forkingMarket = await market.getForkingMarket_();
            const marketDisputeWindow = await market.getDisputeWindow_();
            expect(forkingMarket !== ZERO_ADDRESS || marketDisputeWindow !== disputeWindowAddress).to.be.true;
        }

        if (doGenerateFees) {
            // TODO: Create and call `generateFees` function
        }

        if (moveTimeForward) {
            let disputeWindow = await fixture.getDisputeWindow(market);
            let disputeWindowStartTime = await disputeWindow.getStartTime_();
            await fixture.setTimestamp(disputeWindowStartTime.add(new BN(1)));
        }
    }

    public async proceedToFork(fixture: TestFixture, market: Market) {
        let forkingMarket = await market.getForkingMarket_();
        let disputeRound = 0;
        while (forkingMarket === ZERO_ADDRESS) {
            console.log("\nStarted round", disputeRound);
            await this.proceedToNextRound(fixture, market);
            forkingMarket = await market.getForkingMarket_();
            disputeRound++;
        }

        let ethBalance = await fixture.getEthBalance();
        console.log("ethBalance before calling forkAndRedeem", ethBalance.toString(10));

        const numParticipants = await market.getNumParticipants_();
        for (let i = 0; i < numParticipants.toNumber(); i++) {
            const reportingParticipantAddress = await market.getReportingParticipant_(new BN(i));
            const reportingParticipant = await fixture.getReportingParticipant(reportingParticipantAddress);
            await reportingParticipant.forkAndRedeem();

            const reportingParticipantStake = await reportingParticipant.getStake_();
            expect(reportingParticipantStake === new BN(0));
        }

        ethBalance = await fixture.getEthBalance();
        console.log("ethBalance after calling forkAndRedeem", ethBalance.toString(10));

        console.log("\nCalled forkAndRedeem on reporting participants");
    }
}
